#!/bin/bash

#########################################################
#
# Platform: NCI Gadi HPC
# Description: Runs gatk4_pon_genomicsdbimport.sh in parallel,
# which consolidates per sample pon VCFs per interval in preparation
# for multi-sample calling
# Usage: Adjust <project> and compute resource requests following the guide
# below, then qsub gatk4_pon_genomicsdbimport_run_parallel.pbs
# Job resource requirements:
#       A general guide is 1 hugemem node per ~30 samples for 10Mb interval
# Per task requirements:
# For 10Mb human genomic interval, 34 samples:
#       walltime=00:40:25, average 12-35GB mem, 1 CPU
#       job is memory bound and is influenced by number of variants,
#       cohort size, coverage (which would affect number of variants)
# Author: Tracy Chew
# tracy.chew@sydney.edu.au
# Date last modified: 12/03/2021
#
# If you use this script towards a publication, please acknowledge the
# Sydney Informatics Hub (or co-authorship, where appropriate).
#
# Suggested citation:
# Sydney Informatics Hub, Core Research Facilities, University of Sydney,
# 2021, The Sydney Informatics Hub Bioinformatics Repository, <date accessed>,
# https://github.com/Sydney-Informatics-Hub/Bioinformatics
#
# Suggested acknowledgement:
# The authors acknowledge the scientific and technical assistance
# <or e.g. bioinformatics assistance of <PERSON>> of Sydney Informatics
# Hub and resources and services from the National Computational
# Infrastructure (NCI), which is supported by the Australian Government
# with access facilitated by the University of Sydney.
#
#########################################################

#PBS -P <project>
#PBS -N pon_genomicsdbimport
#PBS -l walltime=02:00:00,ncpus=48,mem=1500GB,wd
#PBS -q hugemem
#PBS -W umask=022
#PBS -l storage=scratch/<project>
#PBS -o ../Logs/pon_genomicsdbimport.o
#PBS -e ../Logs/pon_genomicsdbimport.e

INPUTS=./Inputs/gatk4_pon_genomicsdbimport.inputs

#########################################################
# Do not edit below this line
#########################################################

module load openmpi/4.0.2
module load nci-parallel/1.0.0
module load gatk/4.1.2.0

set -e

# NCPUs = CPUs per task
NCPUS=1

# SCRIPT
SCRIPT=./gatk4_pon_genomicsdbimport.sh

if [[ $PBS_QUEUE =~ bw-exec ]]; then CPN=28; else CPN=48; fi
M=$(( CPN / NCPUS )) #tasks per node

sed "s|^|${SCRIPT} |" ${INPUTS} > ${PBS_JOBFS}/input-file

mpirun --np $((M * PBS_NCPUS / CPN)) \
        --map-by node:PE=${NCPUS} \
        nci-parallel \
        --verbose \
        --input-file ${PBS_JOBFS}/input-file
