#! /bin/bash

#########################################################
#
# Platform: NCI Gadi HPC
# Usage: qsub gatk4_selectvariants.pbs
# Version: 1.0
#
# For more details see: https://github.com/Sydney-Informatics-Hub/Somatic-ShortV
#
# If you use this script towards a publication, support us by citing:
#
# Suggest citation:
# Sydney Informatics Hub, Core Research Facilities, University of Sydney,
# 2021, The Sydney Informatics Hub Bioinformatics Repository, <date accessed>,
# https://github.com/Sydney-Informatics-Hub/Germline-ShortV
#
# Please acknowledge the Sydney Informatics Hub and the facilities:
#
# Suggested acknowledgement:
# The authors acknowledge the technical assistance provided by the Sydney
# Informatics Hub, a Core Research Facility of the University of Sydney
# and the Australian BioCommons which is enabled by NCRIS via Bioplatforms
# Australia. The authors acknowledge the use of the National Computational
# Infrastructure (NCI) supported by the Australian Government.
#
#########################################################

#PBS -P <project>
#PBS -N selectvariants
#PBS -l walltime=04:00:00,ncpus=48,mem=192GB,wd
#PBS -q normal
#PBS -W umask=022
#PBS -l storage=XXX
#PBS -o ./Logs/selectvariants.o
#PBS -e ./Logs/selectvariants.e

logdir=./Logs/gatk4_selectvariants

# Please replace this path with the path to your public variant resource
resource=../Reference/gatk-best-practices/somatic-hg38/af-only-gnomad.hg38.vcf.gz
# Please replace this path with the output path and filename
resource_common=../Reference/gatk-best-practices/somatic-hg38/af-only-gnomad.common_biallelic.hg38.vcf.gz

set -e

module load gatk/4.2.1.0

mkdir -p ${logdir}

gatk --java-options "-Xmx192g" SelectVariants \
        -V ${resource} \
        -select-type SNP -restrict-alleles-to BIALLELIC \
        -select "AF > 0.05" \
        -O ${resource_common} \
        --lenient
