#!/bin/bash

#########################################################
#
# Platform: NCI Gadi HPC
# Description: Gather per interval cohort PoN VCFs. Then use SortVcf
# to create index file and ensure final VCF is coordinate sorted.
# This operates on per interval GenomicsDBImport database files
# Usage: qsub gatk4_cohort_pon_gather_sort.pbs
# Compute resources:
#       Human cohort size of 46 (normal queue):
#       1 CPU, 5.2GB mem, 00:04:52 walltime, 0.76 SUs
# Author: Tracy Chew
# tracy.chew@sydney.edu.au
# Date last modified: 19/03/2021
#
# If you use this script towards a publication, please acknowledge the
# Sydney Informatics Hub (or co-authorship, where appropriate).
#
# Suggested citation:
# Sydney Informatics Hub, Core Research Facilities, University of Sydney,
# 2021, The Sydney Informatics Hub Bioinformatics Repository, <date accessed>,
# https://github.com/Sydney-Informatics-Hub/Bioinformatics
#
# Suggested acknowledgement:
# The authors acknowledge the scientific and technical assistance
# <or e.g. bioinformatics assistance of <PERSON>> of Sydney Informatics
# Hub and resources and services from the National Computational
# Infrastructure (NCI), which is supported by the Australian Government
# with access facilitated by the University of Sydney.
#
#########################################################

#PBS -P er01
#PBS -N cohort_pon_gather_sort
#PBS -l walltime=01:00:00,ncpus=1,mem=10GB,wd
#PBS -q normal
#PBS -W umask=022
#PBS -l storage=scratch/er01+scratch/hm82+scratch/oj47+scratch/public
#PBS -o ./Logs/cohort_pon_gather_sort_samplesSet1andSet2.o
#PBS -e ./Logs/cohort_pon_gather_sort_samplesSet1andSet2.e

set -e

module load gatk/4.1.2.0

cohort=samplesSet1andSet2
config=../$cohort.config
INPUTS=./Inputs
scatterdir=../Reference/ShortV_intervals
scatterlist=$scatterdir/3200_ordered_exclusions.list
vcfdir=./$cohort\_cohort_PoN
logdir=./Logs/gatk4_cohort_pon_gather_sort
args=${INPUTS}/gatk4_cohort_pon_gather_sort.inputs
unsorted=${vcfdir}/${cohort}.pon.vcf.gz
sorted=${vcfdir}/${cohort}.sorted.pon.vcf.gz
num_int=`wc -l ${scatterlist} | cut -d' ' -f 1`

rm -rf ${args}

for interval in $(seq -f "%04g" 0 $((${num_int}-1)));do
        echo "-I" ${vcfdir}/${cohort}.${interval}.pon.vcf.gz >> ${args}
done

mkdir -p ${logdir}
rm -rf ${logdir}/${cohort}.oe

# Gather per interval cohort vcf.gz files into a single .vcf.gz file
# Less KSU to create MD5 later
echo "$(date) : GATK 4 GatherVcfs for CreateSomaticPanelOfNormal calls for $cohort. Gather $cohort.interval.pon.vcf.gz files into $unsorted"
gatk GatherVcfs \
        --arguments_file ${args} \
        --MAX_RECORDS_IN_RAM 100000000 \
        -O ${unsorted} >> ${logdir}/${cohort}.oe 2>&1


# Sort vcf.gz file - in case GatherVcfs fails to make index file (will happen if you output gzipped file)
# Ensures vcf.gz is sorted
echo "$(date) : GATK 4 SortVcf. Soort $unsorted to $sorted"
gatk SortVcf \
        -I ${unsorted} \
        -O ${sorted} >> ${logdir}/${cohort}.oe 2>&1
