#!/bin/bash

#########################################################
#
# Platform: NCI Gadi HPC
# Usage: DO NOT RUN THIS SCRIPT DIRECTLY. It is submitted once for each 
# line in Inputs/gatk4_getpileupsummaries.inputs by wrapper script
# gatk4_getpileupsummaries_byInterval_runLoop.sh
#
# Version: 1.0
#
# For more details see: https://github.com/Sydney-Informatics-Hub/Somatic-ShortV
#
# If you use this script towards a publication, support us by citing:
#
# Suggest citation:
# Sydney Informatics Hub, Core Research Facilities, University of Sydney,
# 2021, The Sydney Informatics Hub Bioinformatics Repository, <date accessed>,
# https://github.com/Sydney-Informatics-Hub/Germline-ShortV
#
# Please acknowledge the Sydney Informatics Hub and the facilities:
#
# Suggested acknowledgement:
# The authors acknowledge the technical assistance provided by the Sydney
# Informatics Hub, a Core Research Facility of the University of Sydney
# and the Australian BioCommons which is enabled by NCRIS via Bioplatforms
# Australia. The authors acknowledge the use of the National Computational
# Infrastructure (NCI) supported by the Australian Government.
#
######################################################### 

#PBS -P xh27
#PBS -l walltime=03:00:00
#PBS -l ncpus=1
#PBS -l mem=31GB
#PBS -l wd
#PBS -q hugemem
#PBS -W umask=022
#PBS -l storage=scratch/er01+gdata/er01+scratch/hm82


set -e

module load gatk/4.2.1.0

ints=($(echo "$intervals" | tr ';' '\n'))

for int in ${ints[@]}
do
	vcf=${vcfdir}/${vcfprefix}/${vcfprefix}.${int}.vcf.gz
	out=${outdir}/${sample}_pileups.${int}.table
	log=${logdir}/${sample}.${int}.log
	
	rm -rf $out $log

	printf "#####################Summarising pileup from $vcf to $out\n"
	gatk GetPileupSummaries \
		--java-options "-Xmx28g -DGATK_STACKTRACE_ON_USER_EXCEPTION=true -XX:ParallelGCThreads=${NCPUS} -Djava.io.tmpdir=${PBS_JOBFS}" \
        	-I ${bam} \
        	-V ${vcf} \
       		-L ${vcf} \
        	-O ${out} > ${log} 2>&1
	
done 
